File: sdk/arena-java-sdk/src/main/java/com/github/kubeflow/arena/client/NodeClient.java
package com.github.kubeflow.arena.client;

import com.alibaba.fastjson.JSON;
import com.github.kubeflow.arena.enums.ArenaErrorEnum;
import com.github.kubeflow.arena.enums.NodeType;
import com.github.kubeflow.arena.exceptions.ArenaException;
import com.github.kubeflow.arena.exceptions.ExitCodeException;
import com.github.kubeflow.arena.model.nodes.*;
import com.github.kubeflow.arena.utils.Command;

import java.io.IOException;
import java.util.List;

public class NodeClient extends BaseClient {

    public NodeClient(String kubeConfig, String namespace, String loglevel, String arenaSystemNamespace) {
        super(kubeConfig, namespace, loglevel, arenaSystemNamespace);
    }

    public NodeClient namespace(String namespace) {
        return new NodeClient(this.kubeConfig, namespace, this.loglevel, this.arenaSystemNamespace);
    }

    public NodeSet all(String... nodeNames) throws ArenaException, IOException {
        return this.filter(NodeType.AllNodeType, nodeNames);
    }

    public NodeSet filter(NodeType nodeType, String... nodeNames) throws ArenaException, IOException {
        List<String> cmds = this.generateCommands("top", "node", "-d", "-o", "json");
        if (!nodeType.equals(NodeType.AllNodeType) && !nodeType.equals(NodeType.UnknownNodeType)) {
            cmds.add("-m=" + nodeType.shortHand());
        }
        if (nodeNames != null && nodeNames.length != 0) {
            for (int i = 0; i < nodeNames.length; i++) {
                cmds.add(nodeNames[i]);
            }
        }
        String[] arenaCommand = cmds.toArray(new String[cmds.size()]);
        try {
            String output = Command.execCommand(arenaCommand);
            NodeSet nodeSet = JSON.parseObject(output, NodeSet.class);
            return nodeSet;
        } catch (ExitCodeException e) {
            throw new ArenaException(ArenaErrorEnum.TOP_NODE, e.getMessage());
        }
    }

    public GPUExclusiveNode[] gpuExclusiveNodes(String... nodeNames) throws ArenaException, IOException {
        NodeSet nodeSet = this.filter(NodeType.GPUExclusiveNodeType, nodeNames);
        return nodeSet.getGpuExclusiveNodes();
    }

    public GPUTopologyNode[] gpuTopologyNodes(String... nodeNames) throws ArenaException, IOException {
        NodeSet nodeSet = this.filter(NodeType.GPUTopologyNodeType, nodeNames);
        return nodeSet.getGpuTopologyNodes();
    }

    public GPUShareNode[] gpuShareNodes(String... nodeNames) throws ArenaException, IOException {
        NodeSet nodeSet = this.filter(NodeType.GPUShareNodeType, nodeNames);
        return nodeSet.getGpuShareNodes();
    }

    public NormalNode[] normalNodes(String... nodeNames) throws ArenaException, IOException {
        NodeSet nodeSet = this.filter(NodeType.NormalNodeType, nodeNames);
        return nodeSet.getNormalNodes();
    }

}

File: sdk/arena-python-sdk/arenasdk/client/node_client.py
#!/usr/bin/env python
from __future__ import annotations
import json
from typing import List
from typing import Dict
from arenasdk.enums.types import *
from arenasdk.nodes.gpu_exclusive_node import GPUExclusiveNode
from arenasdk.nodes.gpu_topology_node  import GPUTopologyNode
from arenasdk.nodes.gpushare_node import GPUShareNode
from arenasdk.nodes.normal_node import NormalNode
from arenasdk.nodes.node_set import NodeSet
from arenasdk.nodes.node_set import build_nodes
from arenasdk.common.util import Command
from arenasdk.exceptions.arena_exception import ArenaException

class NodeClient(object):
    def __init__(self,kubeconfig: str,namespace: str,loglevel: str,arena_namespace: str):
        self.kubeconfig = kubeconfig
        self.namespace = namespace
        self.loglevel = loglevel
        self.arena_namespace = arena_namespace

    def namespace(self,namespace: str) -> TrainingClient:
        return NodeClient(self.kubeconfig,namespace,self.loglevel,self.arena_namespace)
    
    def all(self,*node_names: List[str]) -> NodeSet:
        return self._filter(NodeType.AllNodeType,*node_names)
    
    def gpushare_nodes(self,*node_names: List[str]) -> List[GPUShareNode]:
        return self._filter(NodeType.GPUShareNodeType,*node_names).get_gpushare_nodes()
    
    def gpu_exclusive_nodes(self,*node_names: List[str]) -> List[GPUExclusiveNode]:
        return self._filter(NodeType.GPUExclusiveNodeType,*node_names).get_gpu_exclusive_nodes()
    
    def gpu_topology_nodes(self,*node_names: List[str]) -> List[GPUTopologyNode]:
        return self._filter(NodeType.GPUTopologyNodeType,*node_names).get_gpu_topology_nodes()
    
    def normal_nodes(self,*node_names: List[str]) -> List[NormalNode]:
        return self._filter(NodeType.NormalNodeType,*node_names).get_normal_nodes()

    def _filter(self,node_type: NodeType,*node_names: List[str]) -> NodeSet:
        cmds = self.__generate_commands("top","node","-d","-o","json")
        if node_type != NodeType.AllNodeType and node_type != NodeType.UnknownNodeType:
            cmds.append("-m=" + node_type.value[0])
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                raise ArenaException(ArenaErrorType.TopNodeError,stdout + stderr)
            data = json.loads(stdout)
            return build_nodes(data)
        except ArenaException as e:
            raise e

    def __generate_commands(self,*sub_commands: List[str]) -> List[str]:
        arena_cmds = list()
        arena_cmds.append(ARENA_BINARY)
        for c in sub_commands:
            arena_cmds.append(c)
        if self.kubeconfig != "":
            arena_cmds.append("--config=" + self.kubeconfig)
        if self.namespace != "":
            arena_cmds.append("--namespace=" + self.namespace)
        if self.arena_namespace != "":
            arena_cmds.append("--arena-namespace=" + self.arena_namespace)
        if self.loglevel != "":
            arena_cmds.append("--loglevel=" + self.loglevel)
        return arena_cmds